cli/src/pcluster/resources/custom_resources/custom_resources_code/wait_cluster_ready.py

```diff
@@ -0,0 +1,150 @@
+# Copyright 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
+#  the License. A copy of the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+# OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import string
+
+import boto3
+from botocore.config import Config
+from crhelper import CfnResource
+from utils.ec2_utils import list_cluster_instance_ids_iterator
+from utils.retry_utils import retry
+
+helper = CfnResource(json_logging=False, log_level="INFO", boto_level="ERROR", sleep_on_delete=0)
+logger = logging.getLogger(__name__)
+BOTO3_CONFIG = Config(retries={"max_attempts": 60})
+PAGINATION_CONFIG = {"PageSize": 100}
+
+
+@retry(max_retries=5, wait_time_seconds=3)
+def _get_cluster_config_records(table_name: str, instance_ids: [string]):
+    ddb = boto3.client("dynamodb", config=BOTO3_CONFIG)
+
+    if not instance_ids:
+        logger.warning("No instances to retrieve cluster config records for")
+        return []
+
+    item_ids = [f"CLUSTER_CONFIG.{instance_id}" for instance_id in instance_ids]
+    requested_keys = [{"Id": {"S": item_id}} for item_id in item_ids]
+
+    try:
+        response = ddb.batch_get_item(RequestItems={table_name: {"Keys": requested_keys}})
+        items = response.get("Responses", {}).get(table_name, [])
+    except Exception as e:
+        raise RuntimeError(f"Cannot read config versions due to DDB error: {e}")
+
+    return items
+
+
+def _check_cluster_config_items(instance_ids, items, expected_config_version):
+    # Transform DDB items to make it easier to search
+    items_by_id = {item["Id"]["S"]: item["Data"]["M"] for item in items}
```

Comment:
we can add this translation after the guard for `not instance_ids`  and avoid doing this if we're just returning.